## RNA-seq datasets generation for benchmark

[![N|Solid](https://cldup.com/dTxpPi9lDf.thumb.png)](https://nodesource.com/products/nsolid)
[![Build Status](https://travis-ci.org/joemccann/dillinger.svg?branch=master)](https://travis-ci.org/joemccann/dillinger)

Due to there is no well-studied simulator for noisy long RNA-seq reads, we used PBSIM to generate synthetic datasets by a set of transcripts generated from a particular reference genome and corresponding annotations. Synthetic datasets were created from following genomes and annotations:
  - Homo Sapiens GRCh38 (human) Reference genome, and annotations (version 94) from Ensemble.
  - Mus Musculus GRCm38 (mouse) Reference genome, and annotations (version 94) from Ensemble.
  - Drosophila melanogaster DM6 (fruitfly) genome, and annotations (version 94) from Ensemble.
 ---
RNA-seq simulation datasets were generated by the following workflow:
  1. Download reference genome and annotation file with the same version information. The annotation file should contain the comprehensive gene annotations on the reference chromosomes only. This is the main annotation for most users. Or you can filter gene annotations on scaffolds, assembly patches and alternate loci if there exists.
  2. Extract genes with single splicing isoform, genes with alternative splicing isoforms and genes with short exons (< 30bp). For each kinds of genes, output all the corresponding annotations into three different files by suffix “_AS.gtf”, “_SS.gtf” and “_short.gtf” separately.
  3. Generate transcriptomes from pre-processed three kinds of annotations and reference genome. Filter transcripts short than 200 bps and combine generated transcriptomes together for the input of PBSIM.
  4. For CCS-like, CLR-like, ONT2D-like and ONT1D-like datasets, simulate reads with different error rate (2%, 12%, 15%, 25%), and sequencing depth (4X, 10X, 30X) on the generated transcriptome by PBSIM.
 
### 1. Filter annotations not on chromosomes
As we know, the annotation file download from Ensemble contains all the gene annotations on chromosomes, scaffolds, assemble patches and alternate loci. But many of the scaffolds are unfinshed that should be removed. Or user can download file contains the comprehensive gene annotations on the reference chromosomes only which can be found at GENCODE.

### 2. Extract annotations of different gene types into groups
RNA-seq data is the products of gene transcription, but one gene may transcribe more than one isoforms due to alternative splicing of gene. What's more, an RNA aligner need to not only map exons to genome but also recognize isoforms of genes to reflect the gene structure. Thus, in order to reveal the performance of aligners mapping RNA-seq data to reference genome, we need to simulate RNA-seq data from different aspects to close to real data. In simulation, annotations for each species were seperated into three groups. The first group contains annotations for genes with single splicing isoform, the second group contains annotations for genes with alternative splicing isoforms, the third group contains annotations for genes with short exons (< 30bp). All annotations of each group will be output into files and suffixed by "_SS.gtf", "_AS.gtf" and "_short.gtf" separately. For alternative splicing genes, all isoforms will be extracted. Annotation grouping was down using Annotation_Load.py script in this repository, i.e.

```
python Annotation_grouping.py genome.gtf
```
### 3. Generating transcriptomes
Transcriptomes are generated from processed annotations and genome reference using the script generate_transcriptome.py (city from https://github.com/kkrizanovic/RNAseqEval) in this repository. Since the annotations were separated into three groups, a transcriptom (or a set of transcripts) was generated for each group.
```python
python generate_transcriptome.py SS.gtf ref.fa > SS_transcriptome.fa
python generate_transcriptome.py AS.gtf ref.fa > AS_transcriptome.fa
python generate_transcriptome.py short.gtf ref.fa > short_transcriptome.fa
```
For each transcriptome file, filter transcripts short than 200bp, this was done using fastqfilter.py script from https://github.com/isovic/samscripts with option minlen. And then combine the three transcriptome files together as the input of PBSIM.
```
python fastqfilter min 200 transcriptome.fa > transcriptome_filter.fa
cat SS_transcriptome_filter.fa AS_transcriptome_filter.fa short_transcriptome_filter.fa > merge_transcriptome.fa
```
### 4. Simulation using PBSIM
The merged transcriptome in previouse step was used for simulation. In order to mimic the datasets from various platform, i.e., ONT 1D reads, ONT 2D reads, PacBio subreads (CLR) and PacBio ROI reads (CCS). We simulate four types of reads with 25%, 12%, 15% and 2% error rate respectively for each sepcies. What's more, for each error rate, we simulate dataset with 4X, 10X and 30X sequencing depth to characterize real RNA data. So there are total 36 simulated datasets will be generated by the following command lines:
```
The CCS-like reads:
pbsim 	Transcripome_File (human, mouse, fruitfly)
        --data-type CCS \
        --model_qc model_qc_clr	 \
        --length-mean 6000 \
        --length-min 100
        --difference-ratio 75:5:20 \
        --accuracy-mean 0.98 \
        --accuracy-min 0.8 \
        --depth 4/10/30
```
```
The CLR-like reads: 
	pbsim 	Transcripome_File (human, mouse, fruitfly)
        --data-type CLR \
        --model_qc model_qc_clr	 \
        --length-mean 7800 \
        --length-min 100
        --difference-ratio 1:12:2 \
        --accuracy-mean 0.85 \
        --accuracy-min 0.75 \
        --depth 4/10/30
```
```
The ONT2D-like reads: 
	pbsim 	Transcripome_File (human, mouse, fruitfly)
            --model_qc model_qc_clr	 \
            --length-mean 7800 \
            --length-min 100
            --difference-ratio 33:36:31 \
            --accuracy-mean 0.87 \
            --accuracy-min 0.8 \
            --depth 4/10/30
```

```
The ONT1D-like reads: 
	pbsim 	Transcripome_File (human, mouse, fruitfly)
            --model_qc model_qc_clr	 \
            --length-mean 7800 \
            --length-min 100
            --difference-ratio 48:15:37 \
            --accuracy-mean 0.75 \
            --accuracy-min 0.7 \
            --depth 4/10/30
```
